linear-operators/introduction-to-linear-operators/introduction-to-linear-operators.jmd

```diff
@@ -0,0 +1,302 @@
+$$
+B_{k+1} = B_k + \frac{y_k y_k^T}{y_k^T s_k} - \frac{B_k s_k s_k^T B_k^T}{s_k^T B_k s_k}
+$$
+$$
+H_{k+1} = \bigg(I - \frac{s_k y_k^T}{y_k^T s_k}\bigg)H_k\bigg(I - \frac{y_k s_k^T}{y_k^T s_k}\bigg) + \frac{s_k s_k^T}{y_k^T s_k}
+$$
+
+```julia
+B = LBFGSOperator(5, scaling=false) # B₀ = I
+H = InverseLBFGSOperator(5, scaling=false) # H₀ = I
+
+# Example of s₀ and y₀
+s = rand(5)
+y = rand(5)
+push!(B, s, y)
+push!(H, s, y)
+
+q = rand(5)
+Bq = q + dot(y, q) / dot(y, s) * y - dot(s, q) / dot(s, s) * s
+Hauxq = q - dot(s, q) / dot(y, s) * y
+Hq = Hauxq - dot(y, Hauxq) / dot(y, s) * s + dot(s, q) / dot(y, s) * s
+
+norm(B * q - Bq), norm(H * q - Hq), norm(B * H * q - q)
+```
+
+# Application: Heat Equation
+
+Consider a square plate of size $L \times L$ with no heat exchange on the boundaries.
+The Heat Equation and boundary conditions related to the temperature $u(t,x,y)$ at time $t$, position $(x,y)$ is given by
+$$
+\frac{\partial u}{\partial t} = \alpha \bigg(\frac{\partial^2 u}{\partial x^2} + \frac{\partial^2 u}{\partial y^2}\bigg),
+$$
+where $\alpha > 0$, and
+$$
+\frac{\partial u}{\partial x}(t,0,y) = \frac{\partial u}{\partial x}(t,L,y) =
+\frac{\partial u}{\partial y}(t,x,0) = \frac{\partial u}{\partial y}(t,x,L) = 0.
+$$
+
+Discretizing $u$ into $[0,t_1,\dots,t_N]$ and $x \in [0,L]$ into $[0,x_1,x_2,\dots,x_m,L]$ and accordingly to $y$,
+we define $u^n_{i,j}$ as the approximation to $u(t_n,x_i,y_j)$.
+The derivatives can be approximated by finite differences.
+Using a forward difference for $u_t$ and centered differentes for the spartial differences we obtain
+$$
+\frac{u^{n+1}_{i,j} - u^n_{i,j}}{\delta} = \frac{\alpha}{h^2}(4u^n_{i,j} - u^n_{i-1,j} - u^n_{i+1,j} - u^n_{i,j-1} - u^n_{i,j+1}),
+$$
+where $\delta = t_{n+1} - t_n$ and $h = x_{i+1} - x_i = y_{j+1} - y_j$.
+The boundary conditions lead to
+$$
+u^n_{2,j} = u^n_{0,j}, \quad
+u^n_{m+1,j} = u^n_{m-1,j}, \quad
+u^n_{i,2} = u^n_{i,0}, \quad
+u^n_{i,m+1} = u^n_{i,m-1}.
+$$
+
+```julia
+HeatEquationOp(L, m, δ, α) = HeatEquationOp(Float64, L, m, δ, α)
+
+function HeatEquationOp(opType, L, m, δ, α)
+    h = L / (m + 1)
+    γ = α * δ / h^2
+    κ = 1 + γ
+
+    Tprod(v) = [κ * v[1] - 2γ * v[2];
+             [-γ * v[i-1] + κ * v[i] - γ * v[i+1] for i = 2:m-1];
+              κ * v[m] - 2γ * v[m-1]]
+
+    T = LinearOperator(opType, m, m, false, false, Tprod)
+    D = opEye(m) * γ
+
+    function prod(v)
+        Hv = zeros(m^2)
+        Hv[1:m] .= [T  -2D] * v[1:2m]
+        for i = 2:m-1
+            Hv[(i-1)*m+1:i*m] .= [-D T -D] * v[(i-2)*m+1:(i+1)*m]
+        end
+        Hv[end-m+1:end] .= [-2D T] * v[end-2m+1:end]
+        return Hv
+    end
+
+    return LinearOperator(opType, m^2, m^2, false, false, prod)
+end
+```
+
+```julia
+A = HeatEquationOp(1.0, 3, 0.1, 0.1)
+Matrix(A)
+```
```

Comment:
I don't understand why `A` isn't symmetric. You use central differences for the Laplacian.

Reply:
I tried explaining it better in the new version. Let me know if it's clearer.